{
	"id": "390",
	"name": "Detection of Error Condition Without Action",
	"abstraction": "Base",
	"structure": "Simple",
	"status": "Draft",
	"description": "The product detects a specific error, but takes no actions to handle the error.",
	"related_weaknesses": [
		{
			"nature": "ChildOf",
			"cweid": "755",
			"view_id": "1000",
			"ordinal": "Primary"
		},
		{
			"nature": "CanPrecede",
			"cweid": "401",
			"view_id": "1000"
		}
	],
	"applicable_platforms": {
		"language": [
			{
				"class": "Not Language-Specific",
				"prevalence": "Undetermined"
			}
		]
	},
	"modes_of_introduction": [
		{
			"phase": "Implementation",
			"note": [
				"REALIZATION: This weakness is caused during implementation of an architectural security tactic."
			]
		}
	],
	"likelihood_of_exploit": "Medium",
	"common_consequences": [
		{
			"scope": [
				"Integrity",
				"Other"
			],
			"impact": [
				"Varies by Context",
				"Unexpected State",
				"Alter Execution Logic"
			],
			"note": "An attacker could utilize an ignored error condition to place the system in an unexpected state that could lead to the execution of unintended logic and could cause other unintended behavior."
		}
	],
	"potential_mitigations": [
		{
			"phase": [
				"Implementation"
			],
			"description": [
				"Properly handle each exception. This is the recommended solution. Ensure that all exceptions are handled in such a way that you can be sure of the state of your system at any given moment."
			]
		},
		{
			"phase": [
				"Implementation"
			],
			"description": [
				"If a function returns an error, it is important to either fix the problem and try again, alert the user that an error has happened and let the program continue, or alert the user and close and cleanup the program."
			]
		},
		{
			"phase": [
				"Testing"
			],
			"description": [
				"Subject the product to extensive testing to discover some of the possible instances of where/how errors or return values are not handled. Consider testing techniques such as ad hoc, equivalence partitioning, robustness and fault tolerance, mutation, and fuzzing."
			]
		}
	],
	"demonstrative_examples": [
		{
			"text": "\n               <Intro_Text>The following example attempts to allocate memory for a character. After the call to malloc, an if statement is used to check whether the malloc function failed.</Intro_Text>\n               <Example_Code Nature=\"Bad\" Language=\"C\">\n                  <xhtml:div>foo=malloc(sizeof(char)); //the next line checks to see if malloc failed<xhtml:br/>if (foo==NULL) {<xhtml:div style=\"margin-left:1em;\">//We do nothing so we just ignore the error.</xhtml:div>}</xhtml:div>\n               </Example_Code>\n               <Body_Text>The conditional successfully detects a NULL return value from malloc indicating a failure, however it does not do anything to handle the problem. Unhandled errors may have unexpected results and may cause the program to crash or terminate.</Body_Text>\n               <Body_Text>Instead, the if block should contain statements that either attempt to fix the problem or notify the user that an error has occurred and continue processing or perform some cleanup and gracefully terminate the program. The following example notifies the user that the malloc function did not allocate the required memory resources and returns an error code.</Body_Text>\n               <Example_Code Nature=\"Good\" Language=\"C\">\n                  <xhtml:div>foo=malloc(sizeof(char)); //the next line checks to see if malloc failed<xhtml:br/>if (foo==NULL) {<xhtml:div style=\"margin-left:1em;\">printf(\"Malloc failed to allocate memory resources\");<xhtml:br/>return -1;</xhtml:div>}</xhtml:div>\n               </Example_Code>\n            "
		},
		{
			"text": "\n               <Intro_Text>In the following C++ example the method readFile() will read the file whose name is provided in the input parameter and will return the contents of the file in char string. The method calls open() and read() may result in errors if the file does not exist or does not contain any data to read. These errors will be thrown when the is_open() method and good() method indicate errors opening or reading the file. However, these errors are not handled within the catch statement. Catch statements that do not perform any processing will have unexpected results. In this case an empty char string will be returned, and the file will not be properly closed.</Intro_Text>\n               <Example_Code Nature=\"Bad\" Language=\"C++\">\n                  <xhtml:div>char* readfile (char *filename) {<xhtml:div style=\"margin-left:1em;\">\n                        <xhtml:div>try {<xhtml:div style=\"margin-left:1em;\">\n                              <xhtml:div>// open input file<xhtml:br/>ifstream infile;<xhtml:br/>infile.open(filename);<xhtml:br/>\n                                 <xhtml:br/>if (!infile.is_open()) {<xhtml:div style=\"margin-left:1em;\">throw \"Unable to open file \" + filename;</xhtml:div>}<xhtml:br/>\n                                 <xhtml:br/>// get length of file<xhtml:br/>infile.seekg (0, ios::end);<xhtml:br/>int length = infile.tellg();<xhtml:br/>infile.seekg (0, ios::beg);<xhtml:br/>\n                                 <xhtml:br/>// allocate memory<xhtml:br/>char *buffer = new char [length];<xhtml:br/>\n                                 <xhtml:br/>// read data from file<xhtml:br/>infile.read (buffer,length);<xhtml:br/>\n                                 <xhtml:br/>if (!infile.good()) {<xhtml:div style=\"margin-left:1em;\">throw \"Unable to read from file \" + filename;</xhtml:div>}<xhtml:br/>\n                                 <xhtml:br/>infile.close();<xhtml:br/>\n                                 <xhtml:br/>return buffer;</xhtml:div>\n                           </xhtml:div>}<xhtml:br/>catch (...) {<xhtml:div style=\"margin-left:1em;\">/* bug: insert code to handle this later */</xhtml:div>}</xhtml:div>\n                     </xhtml:div>}</xhtml:div>\n               </Example_Code>\n               <Body_Text>The catch statement should contain statements that either attempt to fix the problem or notify the user that an error has occurred and continue processing or perform some cleanup and gracefully terminate the program. The following C++ example contains two catch statements. The first of these will catch a specific error thrown within the try block, and the second catch statement will catch all other errors from within the catch block. Both catch statements will notify the user that an error has occurred, close the file, and rethrow to the block that called the readFile() method for further handling or possible termination of the program.</Body_Text>\n               <Example_Code Nature=\"Good\" Language=\"C++\">\n                  <xhtml:div>char* readFile (char *filename) {<xhtml:div style=\"margin-left:1em;\">\n                        <xhtml:div>try {<xhtml:div style=\"margin-left:1em;\">\n                              <xhtml:div>// open input file<xhtml:br/>ifstream infile;<xhtml:br/>infile.open(filename);<xhtml:br/>\n                                 <xhtml:br/>if (!infile.is_open()) {<xhtml:div style=\"margin-left:1em;\">throw \"Unable to open file \" + filename;</xhtml:div>}<xhtml:br/>\n                                 <xhtml:br/>// get length of file<xhtml:br/>infile.seekg (0, ios::end);<xhtml:br/>int length = infile.tellg();<xhtml:br/>infile.seekg (0, ios::beg);<xhtml:br/>\n                                 <xhtml:br/>// allocate memory<xhtml:br/>char *buffer = new char [length];<xhtml:br/>\n                                 <xhtml:br/>// read data from file<xhtml:br/>infile.read (buffer,length);<xhtml:br/>\n                                 <xhtml:br/>if (!infile.good()) {<xhtml:div style=\"margin-left:1em;\">throw \"Unable to read from file \" + filename;</xhtml:div>}<xhtml:br/>infile.close();<xhtml:br/>\n                                 <xhtml:br/>return buffer;</xhtml:div>\n                           </xhtml:div>}<xhtml:br/>catch (char *str) {<xhtml:div style=\"margin-left:1em;\">printf(\"Error: %s \\n\", str);<xhtml:br/>infile.close();<xhtml:br/>throw str;</xhtml:div>}<xhtml:br/>catch (...) {<xhtml:div style=\"margin-left:1em;\">printf(\"Error occurred trying to read from file \\n\");<xhtml:br/>infile.close();<xhtml:br/>throw;</xhtml:div>}</xhtml:div>\n                     </xhtml:div>}</xhtml:div>\n               </Example_Code>\n            "
		},
		{
			"text": "\n               <Intro_Text>In the following Java example the method readFile will read the file whose name is provided in the input parameter and will return the contents of the file in a String object. The constructor of the FileReader object and the read method call may throw exceptions and therefore must be within a try/catch block. While the catch statement in this example will catch thrown exceptions in order for the method to compile, no processing is performed to handle the thrown exceptions. Catch statements that do not perform any processing will have unexpected results. In this case, this will result in the return of a null String.</Intro_Text>\n               <Example_Code Nature=\"Bad\" Language=\"Java\">\n                  <xhtml:div>public String readFile(String filename) {<xhtml:div style=\"margin-left:1em;\">\n                        <xhtml:div>String retString = null;<xhtml:br/>try {<xhtml:div style=\"margin-left:1em;\">\n                              <xhtml:div>// initialize File and FileReader objects<xhtml:br/>File file = new File(filename);<xhtml:br/>FileReader fr = new FileReader(file);<xhtml:br/>\n                                 <xhtml:br/>// initialize character buffer<xhtml:br/>long fLen = file.length();<xhtml:br/>char[] cBuf = new char[(int) fLen];<xhtml:br/>\n                                 <xhtml:br/>// read data from file<xhtml:br/>int iRead = fr.read(cBuf, 0, (int) fLen);<xhtml:br/>\n                                 <xhtml:br/>// close file<xhtml:br/>fr.close();<xhtml:br/>\n                                 <xhtml:br/>retString = new String(cBuf);<xhtml:br/>\n                              </xhtml:div>\n                           </xhtml:div>} catch (Exception ex) {<xhtml:div style=\"margin-left:1em;\">/* do nothing, but catch so it'll compile... */</xhtml:div>}<xhtml:br/>return retString;</xhtml:div>\n                     </xhtml:div>}</xhtml:div>\n               </Example_Code>\n               <Body_Text>The catch statement should contain statements that either attempt to fix the problem, notify the user that an exception has been raised and continue processing, or perform some cleanup and gracefully terminate the program. The following Java example contains three catch statements. The first of these will catch the FileNotFoundException that may be thrown by the FileReader constructor called within the try/catch block. The second catch statement will catch the IOException that may be thrown by the read method called within the try/catch block. The third catch statement will catch all other exceptions thrown within the try block. For all catch statements the user is notified that the exception has been thrown and the exception is rethrown to the block that called the readFile() method for further processing or possible termination of the program. Note that with Java it is usually good practice to use the getMessage() method of the exception class to provide more information to the user about the exception raised.</Body_Text>\n               <Example_Code Nature=\"Good\" Language=\"Java\">\n                  <xhtml:div>public String readFile(String filename) throws FileNotFoundException, IOException, Exception {<xhtml:div style=\"margin-left:1em;\">\n                        <xhtml:div>String retString = null;<xhtml:br/>try {<xhtml:div style=\"margin-left:1em;\">\n                              <xhtml:div>// initialize File and FileReader objects<xhtml:br/>File file = new File(filename);<xhtml:br/>FileReader fr = new FileReader(file);<xhtml:br/>\n                                 <xhtml:br/>// initialize character buffer<xhtml:br/>long fLen = file.length();<xhtml:br/>char [] cBuf = new char[(int) fLen];<xhtml:br/>\n                                 <xhtml:br/>// read data from file<xhtml:br/>int iRead = fr.read(cBuf, 0, (int) fLen);<xhtml:br/>\n                                 <xhtml:br/>// close file<xhtml:br/>fr.close();<xhtml:br/>\n                                 <xhtml:br/>retString = new String(cBuf);<xhtml:br/>\n                              </xhtml:div>\n                           </xhtml:div>} catch (FileNotFoundException ex) {<xhtml:div style=\"margin-left:1em;\">System.err.println (\"Error: FileNotFoundException opening the input file: \" + filename );<xhtml:br/>System.err.println (\"\" + ex.getMessage() );<xhtml:br/>throw new FileNotFoundException(ex.getMessage());</xhtml:div>} catch (IOException ex) {<xhtml:div style=\"margin-left:1em;\">System.err.println(\"Error: IOException reading the input file.\\n\" + ex.getMessage() );<xhtml:br/>throw new IOException(ex);</xhtml:div>} catch (Exception ex) {<xhtml:div style=\"margin-left:1em;\">System.err.println(\"Error: Exception reading the input file.\\n\" + ex.getMessage() );<xhtml:br/>throw new Exception(ex);</xhtml:div>}<xhtml:br/>return retString;</xhtml:div>\n                     </xhtml:div>}</xhtml:div>\n               </Example_Code>\n            "
		}
	],
	"observed_examples": [
		{
			"reference": "CVE-2022-21820",
			"description": "A GPU data center manager detects an error due to a malformed request but does not act on it, leading to memory corruption.",
			"link": "https://www.cve.org/CVERecord?id=CVE-2022-21820"
		}
	],
	"references": [
		{
			"reference_id": "REF-18",
			"author": [
				"Secure Software, Inc."
			],
			"title": "The CLASP Application Security Process",
			"url": "https://cwe.mitre.org/documents/sources/TheCLASPApplicationSecurityProcess.pdf",
			"publication_year": "2005"
		},
		{
			"section": "\"Sin 11: Failure to Handle Errors Correctly.\" Page 183",
			"reference_id": "REF-44",
			"author": [
				"Michael Howard",
				"David LeBlanc",
				"John Viega"
			],
			"title": "24 Deadly Sins of Software Security",
			"publication_year": "2010",
			"publication": "McGraw-Hill"
		}
	],
	"content_history": {
		"submission": {
			"submission_name": "CLASP",
			"submission_date": "2006-07-19"
		},
		"modification": [
			{
				"modification_name": "Eric Dalci",
				"modification_organization": "Cigital",
				"modification_date": "2008-07-01",
				"modification_comment": "updated Time_of_Introduction"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2008-09-08",
				"modification_comment": "updated Relationships, Other_Notes, Taxonomy_Mappings"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2008-11-24",
				"modification_comment": "updated Demonstrative_Examples, Description, Other_Notes, Potential_Mitigations"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2009-03-10",
				"modification_comment": "updated Relationships"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2009-07-27",
				"modification_comment": "updated Demonstrative_Examples"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2011-06-01",
				"modification_comment": "updated Common_Consequences, Relationships, Taxonomy_Mappings"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2011-06-27",
				"modification_comment": "updated Common_Consequences"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2011-09-13",
				"modification_comment": "updated Relationships, Taxonomy_Mappings"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2012-05-11",
				"modification_comment": "updated Common_Consequences, References, Relationships"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2014-02-18",
				"modification_comment": "updated Related_Attack_Patterns"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2014-07-30",
				"modification_comment": "updated Relationships, Taxonomy_Mappings"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2017-11-08",
				"modification_comment": "updated Applicable_Platforms, Modes_of_Introduction, Relationships, Taxonomy_Mappings"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2019-01-03",
				"modification_comment": "updated Related_Attack_Patterns, Taxonomy_Mappings"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2020-02-24",
				"modification_comment": "updated References, Type"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2020-08-20",
				"modification_comment": "updated Relationships"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2023-01-31",
				"modification_comment": "updated Description, Potential_Mitigations"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2023-04-27",
				"modification_comment": "updated Detection_Factors, Relationships, Time_of_Introduction"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2023-06-29",
				"modification_comment": "updated Mapping_Notes"
			},
			{
				"modification_name": "CWE Content Team",
				"modification_organization": "MITRE",
				"modification_date": "2023-10-26",
				"modification_comment": "updated Observed_Examples"
			}
		],
		"previous_entry_name": [
			{
				"text": "Improper Error Handling",
				"date": "2008-04-11"
			}
		]
	},
	"detection_methods": [
		{
			"detection_method_id": "DM-14",
			"method": "Automated Static Analysis",
			"description": "Automated static analysis, commonly referred to as Static Application Security Testing (SAST), can find some instances of this weakness by analyzing source code (or binary/compiled code) without having to execute it. Typically, this is done by building a model of data flow and control flow, then searching for potentially-vulnerable patterns that connect \"sources\" (origins of input) with \"sinks\" (destinations where the data interacts with external components, a lower layer such as the OS, etc.)",
			"effectiveness": "High"
		}
	],
	"taxonomy_mappings": [
		{
			"taxonomy_name": "CLASP",
			"entry_name": "Improper error handling"
		},
		{
			"taxonomy_name": "The CERT Oracle Secure Coding Standard for Java (2011)",
			"entry_name": "Do not suppress or ignore checked exceptions",
			"entry_id": "ERR00-J"
		},
		{
			"taxonomy_name": "Software Fault Patterns",
			"entry_name": "Unchecked Status Condition",
			"entry_id": "SFP4"
		}
	]
}
